### PR TITLE
Remove Regex-Replaces/Raw String ops from #2660 (AssignedByValParameterQuickFix)

### PR DESCRIPTION
**Pull request by BZngr**
Updates AssignedByValParameterQuickFix.cs to eliminate use of Regular Expressions in favor of Parse Tree driven updates.  Refactored code/logic duplicated within AssignedByValParameterQuickFix.cs and PassParameterByReferenceQuickFix.cs to new class AssignedByValParameterQuickFixHelper.cs.

### CODE REVIEW

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs [hunk 1]
```diff
@@ -19,25 +17,29 @@ public class AssignedByValParameterQuickFix : QuickFixBase
         private readonly Declaration _target;
         private string _localCopyVariableName;
         private bool _isQuickFixUnitTest;
-        private string[] _originalProcCodeLines;
+        private AssignedByValParameterQuickFixHelper _quickFixHlpr;
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: why the name-shortening? `_quickFixHelper` seems to work just as well...

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs [hunk 2]
```diff
@@ -70,75 +75,46 @@ private void SetLocalCopyVariableName()
             }
         }
 
-        private void ModifyBlockToUseLocalCopyVariable()
+        private bool ProposedLocalVariableNameIsValid()
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: might be cleaner to just pass the name into this method here. This would allow extracting the validity-check into a separate class, instead of reinventing that wheel over and over across the codebase.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs [hunk 3]
```diff
@@ -151,36 +127,5 @@ private string BuildLocalCopyAssignment()
             return (SymbolList.ValueTypes.Contains(_target.AsTypeName) ? string.Empty : Tokens.Set + " ") 
                 + _localCopyVariableName + " = " + _target.IdentifierName;
         }
-        
-        private string[] GetAllModuleLines()
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: glad to see this and `GetProcedureLines` gone from here...

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+using System.Diagnostics;
+using System.Linq;
+
+namespace Rubberduck.Inspections.QuickFixes
+{
+    class AssignedByValParameterQuickFixHelper
+    {
+        private Declaration _target;
+        private QualifiedSelection _selection;
+        private HashSet<string> _namesInUse;
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Type seems overspecified here. What's wrong with `Set`?

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        private ICodeModule _module;
+        public AssignedByValParameterQuickFixHelper(Declaration target, QualifiedSelection selection)
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: The name suggests that this is only usable for `AssignedByValParameterQuickFix`.. is that true? If yes, why?

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            }
+            else if (context is VBAParser.PropertyGetStmtContext)
+            {
+                return ((VBAParser.PropertyGetStmtContext)context).argList().arg();
+            }
+            else if (context is VBAParser.PropertySetStmtContext)
+            {
+                return ((VBAParser.PropertySetStmtContext)context).argList().arg();
+            }
+            return null;
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: empty list seems to be preferrable. It allows us to not have to null-check

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            }
+            else if (context is VBAParser.PropertyGetStmtContext)
+            {
+                return ((VBAParser.PropertyGetStmtContext)context).block().blockStmt();
+            }
+            else if (context is VBAParser.PropertySetStmtContext)
+            {
+                return ((VBAParser.PropertySetStmtContext)context).block().blockStmt();
+            }
+            return null;
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Empty list seems preferrable

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            {
+                return ((VBAParser.PropertyGetStmtContext)context).argList().arg();
+            }
+            else if (context is VBAParser.PropertySetStmtContext)
+            {
+                return ((VBAParser.PropertySetStmtContext)context).argList().arg();
+            }
+            return null;
+        }
+        private HashSet<string> GetVariableNamesUsedInProcedureContext()
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: again: type seems overspecified. 
Also: Why not access the parser-state here and query the Declarations?

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs [hunk 9]
```diff
@@ -15,11 +15,13 @@ namespace Rubberduck.Inspections.QuickFixes
     public class PassParameterByReferenceQuickFix : QuickFixBase
     {
         private Declaration _target;
+        private AssignedByValParameterQuickFixHelper _quickFixHlpr;
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: why is the `AssignedByValParameterQuickFixHelper` used in `PassParameterByReferenceQuickFix`?? This seems incorrect. Probably the best workaround would be to name it `QuickFixHelper`, largely because it seems completely unrelated to any special quick-fix as of now..

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs [hunk 10]
```diff
@@ -30,59 +32,19 @@ public override void Fix()
...
-            ReplaceModuleLine(terminalNode.Symbol.Line, replacementLine);
+            _quickFixHlpr.ReplaceModuleLine(terminalNode.Symbol.Line, replacementLine);
 
         }
         private VBAParser.ArgContext GetArgContextForIdentifier(RuleContext context, string identifier)
         {
-            var argList = GetArgListForContext(context);
-            return argList.arg().SingleOrDefault(parameter =>
+            var args = _quickFixHlpr.GetArgContextsForContext(context);
+            return args.SingleOrDefault(parameter =>
```
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: I'd *personally* inline `args` here, but YMMV

File: RetailCoder.VBE/UI/Refactorings/AssignedByValParameterQuickFixDialog.cs [hunk 11]
```diff
@@ -1,22 +1,22 @@
...
 
-        public AssignedByValParameterQuickFixDialog(string[] procedureLines)
+        public AssignedByValParameterQuickFixDialog(Declaration target, QualifiedSelection selection)
         {
-            _procedureLines = procedureLines;
-            _userInputIsValid = false;
             InitializeComponent();
             InitializeCaptions();
+            Target = target;
+            _quickFixHlpr = new AssignedByValParameterQuickFixHelper(target, selection);
```
Comment (Vogel612) [inline RetailCoder.VBE/UI/Refactorings/AssignedByValParameterQuickFixDialog.cs]: why not reuse the helper instance? Why does the dialog need the helper instance in the first place? the decoupling of presentation and business logic seems to lack a bit here :)

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs (hunk 1 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: Will implement

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs (hunk 2 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: There is already a VariableNameValidator class (used internally in this QuickFix as well) that has the generic checks for variable name validity - only to the scope of the variable name as a standalone name.  This function would resist extraction a bit in that it has an additional check beyond the VariableNameValidator class (e.g., the variable name is already used somewhere in the procedure).

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 4 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: I'm new to C#, but my initial checks online and simply typing "Set" into VS indicate that Set does not exist in C#. (?) HashSet was selected to create a list/set of unique values.  If I'm mistaken about "Set", I am happy to change this.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 5 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: My hope is that there may be some functionality in this class that is useful on a more general basis.  I'm a new contributor, so I am exceedingly hesitant to propose a system-wide helper class.  So, I intentionally named this class specific to the QuickFix(es) it supports...That said, if a class with similar functionality does not already exist, and other (more experienced) contributors believe there is merit to making some or all of this class more general...that would be great.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 7 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Agreed.  Will implement

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 6 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Agreed. Will implement

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 8 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: I've yet to find a 'quick' way to get all the declarations in both the parameter list and the procedure body (does it exist?).  I'm looking for all possible parameters/declarations in the function to compare against a name entered by the user - so he can be made aware of any collision(s).

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 9 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: I've used the helper class within PassParameterByReferenceQuickFix because this QuickFix is only used by the AssignedByValParameterInspection.  Another alternative is to rename the two related QuickFix classes to: AssignedByValParameterMakeLocalCopyQuickFix and AssignedByValParameterChangeToByRefQuickFix.  This would be more precise, and then the helper class name makes a little more sense.  Going to the more generic QuickFixHelper is also OK, but it implies that the class is will need to meet a bit higher standard as generic helper.

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 10 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: Will implement - left it separated initially for debug/development.

File: RetailCoder.VBE/UI/Refactorings/AssignedByValParameterQuickFixDialog.cs (hunk 11 above)
Comment (BZngr) [inline RetailCoder.VBE/UI/Refactorings/AssignedByValParameterQuickFixDialog.cs]: The helper is used in the dialog for the sole purpose of supporting dynamic evaluation of the user-entered variable name.  So, rather than have the dialog call back to the QuickFix class directly for this validation support, I chose to have it use the helper class.  Another option is to pass the dialog the list of 'forbidden' names (in the ctor or as a property) and remove the helper altogether (I'm thinking I like the last option best).

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 4 above)
Comment (Vogel612) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: my C# is a little rusty, too ... I was not aware ..

Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: `Set<T>` is Java ;-)

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 7 above)
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: That would be `Enumerable.Empty<VBAParser.BlockStmtContext>()`

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 8 above)
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: The `DeclarationFinder` has a `FindParameter(Declaration procedure, string parameterName)` method that returns one specific parameter declaration (or `null`):

````csharp
public Declaration FindParameter(Declaration procedure, string parameterName)
{
    return _parametersByParent[procedure]
        .SingleOrDefault(parameter => parameter.IdentifierName == parameterName);
}
````

I think the single most efficient way to return all parameters is to leverage the finder's `_parametersByParent` dictionary. It could look like this:

````csharp
public IEnumerable<Declaration> GetProcedureParameters(Declaration procedure)
{
    return _parametersByParent[procedure].ToList();
}
````

What we're really missing is a way to instantly/efficiently get all *accessible* declarations given a *scope*: checking only under the scope of the procedure we're looking at isn't sufficient, because a variable could be declared at module level, or in global scope, with the same name too, and naming a local/parameter with that identifier would "hide" it.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+            foreach (RuleContext ruleContext in ruleContexts)
+            {
+                var identifiersForThisContext = GetIdentifierNames(ruleContext);
+                identifiers.UnionWith(identifiersForThisContext);
+            }
+            return identifiers;
+        }
+        private HashSet<string> GetIdentifierNames(RuleContext ruleContext)
+        {
+            //Recursively work through the tree to get all IdentifierContexts
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Makes me wonder if a parse tree listener wouldn't be simpler?

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 9 above)
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: I like the idea (very much!) of moving code out of quickfix implementations and into a common "helper" class, however it needs a more specialized name than "helper" (I'm thinking `CodePaneRewriter` or something); know however that we are planning to come up with an actual API specifically for this (ref. #2688), so I think I'd rather have efforts concentrated more toward shaping that API than refactoring things into a temporary helper that will end up superseded by that API - although, that helper can help us figure out everything this API needs to be able to do.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs [hunk 13]
```diff
@@ -47,17 +49,20 @@ public override void Fix()
...
+                return;
+            }
 
             ModifyBlockToUseLocalCopyVariable();
         }
 
         private void SetLocalCopyVariableName()
         {
-            using (var view = new AssignedByValParameterQuickFixDialog(_originalProcCodeLines))
+            using (var view = new AssignedByValParameterQuickFixDialog(_target, Selection))
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: This hard-wired dependency is going to be a problem with unit tests: we simply cannot have a UI pop up in the middle of an automated test suite running on a CI server.

Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFix.cs]: For this exact reason, the AssignedByValParameterQuickFix has an accessor 'TESTONLY_FixUsingAutoGeneratedName(...)' [not on the QuickFix interface].  This function is there for the sole purpose of supporting unit tests.  This accessor calls "Fix()" after setting an internal flag that suppresses the call to 'ShowDialog()'.  Didn't really like adding the test-support method, but couldn't think of (or find an example) of a better way.

File: RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs (hunk 7 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: Using Enumerable.Empty seems to require a lengthy cast to compile.  For brevity,  Is ` return new List<VBAParser.BlockStmtContext>(); ` equivalent?``

Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/AssignedByValParameterQuickFixHelper.cs]: That would be because of the `IReadOnlyList` I think. `Enumerable.Empty<VBAParser.BlockStmtContext>().ToArray()` would do it.

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 9 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: So...I'm going to change the names of the two AssignedByVal* QuickFIxes to be more precise.  I'll also change the helper class name to the more generic `QuickFixHelper.`  Over time, the hope is that the helper class disappears in favor of a future `CodPaneRewriter` class.

**Comment by BZngr** [comment]
Addressed comments from the initial pull request.  Created test class for each AssignedByXXXValQuickFix as the original test class had become too large and unwieldy.  Only inspection tests remain in the initial test class.  Moved the IgnoreOnceQuickFix test into the existing IgnoreQuickFixTests class.

**Comment by BZngr** [comment]
Not sure how to resolve these conflicts.  PassParameterByReferenceQuickFix.cs is deleted by this PR - and is no longer in my local project.  Neither the above file, nor AssignedByValParameterInspectionTests.cs is updated when I do a git merge - the merge command indicates I have the latest.  (?)

File: RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs [hunk 14]
```diff
@@ -10,28 +10,32 @@
 
 namespace Rubberduck.Inspections.QuickFixes
 {
-    class AssignedByValParameterQuickFixHelper
+    class QuickFixHelper
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: I think, because its tightly coupled with the `FoobarQuickFix` implementations - and IMO that's a fine coupling to have here (although a few tests for it would be golden), so we should make the class `static` and remove all state: the `target` and `selection` become parameters of the members that need them, and the members become `static` as well, and then some could also be implemented as extension methods that extend `ICodePane`, for example.

File: RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs [hunk 15]
```diff
@@ -10,28 +10,32 @@
...
-            _namesInUse = new HashSet<string>();
+            _identifierNamesDeclaredInProcedureScope = new HashSet<string>();
             _module = selection.QualifiedName.Component.CodeModule;
         }
-        public string GetModuleLine(int line)
+        /// <summary>
+        /// Returns the CodeModule string located at lineNumber
+        /// of the newly inserted line.
+        /// </summary>
+        public string GetModuleLine(int lineNumber)
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: (continued) as an extension method that would be `public static string GetModuleLine(this ICodeModule module, int line)` - with it we could call `GetModuleLine` straight from any `ICodeModule` instance.

File: RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs [hunk 16]
```diff
@@ -43,41 +47,36 @@ public int InsertModuleLinesAfter(int lineNumber, string[] newContent)
...
-            return line;
+            _module.ReplaceLine(lineNumber, newContent);
+            return lineNumber;
         }
-        public string ReplaceStringAtIndex(int lineNumber, string toReplace, string replacement, int startIndex)
+        /// <summary>
+        /// Replaces a the TerminalNode Text value found at the location specified by the 
+        /// TerminalNode context.
+        /// </summary>
+        public void ReplaceTerminalNodeTextInCodeModule(ITerminalNode terminalNode, string replacement)
```
Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: Here would be `public static void ReplaceToken(this ICodeModule module, IToken token, string replacement)` ...and we definitely *need* that one!!!

File: RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs (hunk 15 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: That all sounds good to me.  I'll make those types of changes on the next pass.  Thanks.

File: RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs (hunk 16 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: Thanks - I would not have known if "Replace Token" would be sufficiently clear to other contributors.  Shorter names are always better if no meaning is lost.

Comment (retailcoder) [inline RetailCoder.VBE/Inspections/QuickFixes/QuickFixHelper.cs]: As an extension method to `ICodeModule`, it is - as a member of a "Helper" class, not much 😄

File: RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs (hunk 9 above)
Comment (BZngr) [inline RetailCoder.VBE/Inspections/QuickFixes/PassParameterByReferenceQuickFix.cs]: In the follow-on to this PR (PR #2733) I elected to keep the PassParameterByRefernceQuickFIx named as-is.  Even though it is currently only used by the AssignedByValParameter inspection, it looks as though it would be useful for any inspection that has a need to change a variable from ByVal to ByRef.

**Comment by BZngr** [comment]
This PR and discussion threads are continued at PR #2733.
